### PR TITLE
Refactor shadow algorithm.

### PR DESCRIPTION
**Pull request by soulasuna**
For #11661.
Refactor shadow algorithm.

Changes proposed in this pull request:
- Add algorithm secondary interface.
- Add check on shadow Algorithm.
- Add  shadow Algorithm method.

### CODE REVIEW

File: shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-api/src/main/java/org/apache/shardingsphere/shadow/api/shadow/column/ShadowOperationType.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+     */
+    SELECT;
+    
+    /**
+     * Create shadow operation type.
+     *
+     * @param operationType operation type
+     * @return shadow operation type
+     */
+    public static Optional<ShadowOperationType> newInstance(final String operationType) {
```
Comment (tristaZero) [inline shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-api/src/main/java/org/apache/shardingsphere/shadow/api/shadow/column/ShadowOperationType.java:56]: Could your refer to `MySQLAuthenticationMethod`?

File: shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-core/src/main/java/org/apache/shardingsphere/shadow/algorithm/shadow/ShadowAlgorithmType.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+     */
+    SIMPLE_NOTE;
+    
+    /**
+     * Create shadow algorithm type.
+     *
+     * @param shadowAlgorithmType shadow algorithm type
+     * @return shadow operation type
+     */
+    public static Optional<ShadowAlgorithmType> newInstance(final String shadowAlgorithmType) {
```
Comment (tristaZero) [inline shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-core/src/main/java/org/apache/shardingsphere/shadow/algorithm/shadow/ShadowAlgorithmType.java:43]: Do you think `instanceof` is suitable to replace this class?

File: shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-core/src/main/java/org/apache/shardingsphere/shadow/algorithm/shadow/note/SimpleSQLNoteShadowAlgorithm.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    @Override
+    public boolean isShadow(final Collection<String> shadowTableNames, final PreciseNoteShadowValue<String> noteShadowValue) {
+        if (!shadowTableNames.contains(noteShadowValue.getLogicTableName())) {
+            return false;
+        }
+        Optional<Map<String, String>> noteOptional = parseNote(noteShadowValue.getSqlNoteValue());
+        return noteOptional.filter(stringStringMap -> props.entrySet().stream().allMatch(entry -> Objects.equals(entry.getValue(), stringStringMap.get(String.valueOf(entry.getKey()))))).isPresent();
+    }
+    
+    private Optional<Map<String, String>> parseNote(final String sqlNoteValue) {
```
Comment (tristaZero) [inline shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-core/src/main/java/org/apache/shardingsphere/shadow/algorithm/shadow/note/SimpleSQLNoteShadowAlgorithm.java:73]: It is suggested to extract this logic to a util class for other `note algorithms`.

File: shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-api/src/main/java/org/apache/shardingsphere/shadow/api/shadow/note/PreciseNoteShadowValue.java [hunk 4]
```diff
@@ -15,24 +15,19 @@
...
 
-/**
- * Simple note shadow algorithm.
- */
+@RequiredArgsConstructor
 @Getter
-@Setter
-public final class SimpleSQLNoteShadowAlgorithm implements ShadowAlgorithm {
+@ToString
+public class PreciseNoteShadowValue<T extends Comparable<?>> implements ShadowValue {
```
Comment (tristaZero) [inline shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-api/src/main/java/org/apache/shardingsphere/shadow/api/shadow/note/PreciseNoteShadowValue.java:28]: Is it supposed to be a final class?

File: shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-api/src/main/java/org/apache/shardingsphere/shadow/api/shadow/column/PreciseColumnShadowValue.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import lombok.ToString;
+import org.apache.shardingsphere.shadow.api.shadow.ShadowValue;
+
+/**
+ * Shadow value for precise column.
+ */
+@RequiredArgsConstructor
+@Getter
+@ToString
+public class PreciseColumnShadowValue<T extends Comparable<?>> implements ShadowValue {
```
Comment (tristaZero) [inline shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-api/src/main/java/org/apache/shardingsphere/shadow/api/shadow/column/PreciseColumnShadowValue.java:31]: Is it supposed to be a final class?

**Review by tristaZero — CHANGES_REQUESTED** [review]

File: shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-core/src/main/java/org/apache/shardingsphere/shadow/algorithm/shadow/ShadowAlgorithmType.java (hunk 2 above)
Comment (soulasuna) [reply shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-core/src/main/java/org/apache/shardingsphere/shadow/algorithm/shadow/ShadowAlgorithmType.java:43]: OK. This class will be removed.

File: shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-core/src/main/java/org/apache/shardingsphere/shadow/algorithm/shadow/note/SimpleSQLNoteShadowAlgorithm.java (hunk 3 above)
Comment (soulasuna) [reply shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-core/src/main/java/org/apache/shardingsphere/shadow/algorithm/shadow/note/SimpleSQLNoteShadowAlgorithm.java:73]: I will fix it.

File: shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-api/src/main/java/org/apache/shardingsphere/shadow/api/shadow/note/PreciseNoteShadowValue.java (hunk 4 above)
Comment (soulasuna) [reply shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-api/src/main/java/org/apache/shardingsphere/shadow/api/shadow/note/PreciseNoteShadowValue.java:28]: I will fix it.

File: shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-api/src/main/java/org/apache/shardingsphere/shadow/api/shadow/column/PreciseColumnShadowValue.java (hunk 5 above)
Comment (soulasuna) [reply shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-api/src/main/java/org/apache/shardingsphere/shadow/api/shadow/column/PreciseColumnShadowValue.java:31]: I will fix it.

File: shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-api/src/main/java/org/apache/shardingsphere/shadow/api/shadow/column/ShadowOperationType.java (hunk 1 above)
Comment (soulasuna) [reply shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-api/src/main/java/org/apache/shardingsphere/shadow/api/shadow/column/ShadowOperationType.java]: OK.I will optimization method name.

**Review by tristaZero — APPROVED** [review]

**Pull request merged**